src/registry.rs

```diff
@@ -0,0 +1,238 @@
+    /// let (ret, path) = registry.resolve(&url).unwrap();
+    /// assert_eq!(path.as_ref(), "path/to/object");
+    /// assert!(Arc::ptr_eq(&ret, &bucket2));
+    ///
+    /// let bucket3 = Arc::new(PrefixStore::new(InMemory::new(), "path")) as Arc<dyn ObjectStore>;
+    /// let base = Url::parse("https://s3.region.amazonaws.com/bucket/path").unwrap();
+    /// registry.register(base, bucket3.clone());
+    ///
+    /// let url = Url::parse("https://s3.region.amazonaws.com/bucket/path/to/object").unwrap();
+    /// let (ret, path) = registry.resolve(&url).unwrap();
+    /// assert_eq!(path.as_ref(), "to/object");
+    /// assert!(Arc::ptr_eq(&ret, &bucket3));
+    /// ```
+    fn resolve(&self, url: &Url) -> crate::Result<(Arc<dyn ObjectStore>, Path)>;
+}
+
+/// An [`ObjectStoreRegistry`] that uses [`parse_url`] to create stores based on the environment
+#[derive(Debug, Default)]
+pub struct DefaultObjectStoreRegistry {
+    map: RwLock<BTreeMap<StrUrl, RegistryEntry>>,
+}
+
+/// Wrapper around [`Url`] implementing `Borrow<str>`
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq)]
+struct StrUrl(Url);
+
+impl Borrow<str> for StrUrl {
+    fn borrow(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+#[derive(Debug)]
+struct RegistryEntry {
+    store: Arc<dyn ObjectStore>,
+    /// The number of non-empty path segments in the store's URL
+    path_segments: usize,
+}
+
+impl RegistryEntry {
+    fn new(url: &Url, store: Arc<dyn ObjectStore>) -> Self {
+        let path_segments = match url.path() {
+            "/" => 0,
+            path => path.as_bytes().iter().filter(|&&c| c == b'/').count(),
+        };
+        Self {
+            path_segments,
+            store,
+        }
+    }
+
+    fn resolve(&self, url: &Url) -> Result<(Arc<dyn ObjectStore>, Path), Error> {
+        let segments = match url.path_segments() {
+            Some(segments) => segments.skip(self.path_segments),
+            None => return Ok((self.store.clone(), Path::default())),
+        };
+
+        let path = segments.map(PathPart::parse).collect::<Result<_, _>>()?;
+        Ok((self.store.clone(), path))
+    }
+}
+
+impl DefaultObjectStoreRegistry {
+    /// Create a new [`DefaultObjectStoreRegistry`]
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl ObjectStoreRegistry for DefaultObjectStoreRegistry {
+    fn register(&self, url: Url, store: Arc<dyn ObjectStore>) -> Option<Arc<dyn ObjectStore>> {
+        let entry = RegistryEntry::new(&url, store);
+        Some(self.map.write().insert(StrUrl(url), entry)?.store)
+    }
+
+    fn unregister(&self, store: &Url) -> Option<Arc<dyn ObjectStore>> {
+        Some(self.map.write().remove(store.as_str())?.store)
+    }
+
+    fn resolve(&self, url: &Url) -> crate::Result<(Arc<dyn ObjectStore>, Path)> {
+        {
+            let map = self.map.read();
+            let start = Bound::Included(&url[..url::Position::BeforePath]);
+
+            let longest = map
```

Comment:
This is the downside of this formulation, we have to use a BTreeMap which is slower than a HashMap, and if there are multiple (scheme, host) matches we have to scan through them